rfc/0015-multi-file-jayvee/0015-multi-file-jayvee.md

```diff
@@ -0,0 +1,160 @@
+<!--
+SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+
+SPDX-License-Identifier: AGPL-3.0-only
+-->
+
+# RFC 0015: Multi-file Jayvee
+
+|             |                 |
+| ----------- | --------------- | --------------------------------------------------------------- |
+| Feature Tag | `multi-file`    |
+| Status      | `DRAFT`         | <!-- Possible values: DRAFT, DISCUSSION, ACCEPTED, REJECTED --> |
+| Responsible | `georg-schwarz` | <!-- TODO: assign yourself as main driver of this RFC -->       |
+
+<!--
+  Status Overview:
+  - DRAFT: The RFC is not ready for a review and currently under change. Feel free to already ask for feedback on the structure and contents at this stage.
+  - DISCUSSION: The RFC is open for discussion. Usually, we open a PR to trigger discussions.
+  - ACCEPTED: The RFC was accepted. Create issues to prepare implementation of the RFC.
+  - REJECTED: The RFC was rejected. If another revision emerges, switch to status DRAFT.
+-->
+
+## Summary
+
+This RFC introduces the possibility of distributing a Jayvee program over multiple files.
+This feature will foster reuse of valuetypes, blocks, and other elements.
+Inherent to this feature is a concept of how scoping and naming is handled for nested structures.
+We introduce two concepts in this RFC:
+
+- File imports, and
+- Libraries
+
+## Motivation
+
+Right now, Jayvee users can only pack their whole Jayvee model into one file.
+The challenge is two-fold:
+
+1. Larger projects will become unmaintainable quite quickly, as the elements cannot be organized into multiple files.
+2. Without distribution to multiple files, there is no possibility to reuse models of other projects.
+
+For example, we might be enable to build libraries of valuetypes that can be reused across multiple projects instead of copying the code.
+
+## Explanation
+
+### Exporting elements for later import
+
+For exporting single elements, I propose to introduce a new keyword `export`.
+All elements within a file are not exportable per default.
+Explicitly declaring an element as exportable allows for later import.
+
+**Example export**
+
+```
+export valuetype MyValueType {
+  // ... details
+}
+```
+
+### Bundling elements to a library for later import
+
+For bundling and exporting elements, I propose introducing a new concept called `libraries`.
+A `library` can inhibit `Valuetype`s, `Block`s, `BlockType`s, `Constraint`s, and `Transform`s.
+A library has to define a version in semver syntax.
+A library has to be exported.
+They serve as entry point to a collection of files (local or remote), similarly to JavaScript's `index.js` mechanism.
+
+**Example library**
+
+```
+export library MyDomainLibrary version 1.2.3 {
+  // definition of a new valuetype as part of the library
+  valuetype MyDomainSpecificValuetype1 {
+    // ... details of valuetype
+  }
+
+  // reference to an existing valuetype to make it part of the library
+  include MyDomainSpecificValuetype2;
```

Comment:
I still dislike this one. Any ideas for a better syntax?

An alternative could be disallowing defining new elements in a library and only use it as an export list (potentially with the `called` keyword)

Reply:
I don't understand why you need to include things here? Why not just require things to be defined in the lib?

Reply:
My thoughts were:
- if a library is the "entry point" to reusing a jv project, it might make sense to decouple it from the internal structure of the project
- For example, the user might define their valuetypes in a `my-valuetypes.jv`, but still wants to publish them with the library
- Otherwise, the library definition might become very large because it has to define everything that is accessible by other projects

Reply:
Changed it to `publish <ElementName>;` as an alternative to definition within the library.